conda_mirror/conda_mirror.py

```diff
@@ -238,161 +317,105 @@ def main(upstream_channel, target_directory, platform, blacklist=None,
-        e.g., this is the folder that should contain all of the conda packages
-        and a "repodata.json"
-
-    Yields
-    ------
-    full_pkg_path : str
-        The full path to a package that results in `conda_build.index:read_index_tar()`
-        raising a tarfile.ReadError
-    """
-    repodata_fname = os.path.join(local_platform_directory, 'repodata.json')
-    repodata = {'info': {}, 'packages': {}}
-    if os.path.exists(repodata_fname):
-        with open(repodata_fname, 'r') as f:
-            repodata = json.load(f)
-    repodata_info, repodata_packages = repodata.get('info', {}), repodata.get('packages', {})
-    indexed_packages = list(repodata_packages.keys())
-    all_packages = fnmatch.filter(os.listdir(local_platform_directory), "*.tar.bz2")
-    potentially_bad = set(all_packages).difference(indexed_packages)
-    for pkg in potentially_bad:
-        full_pkg_path = os.path.join(local_platform_directory, pkg)
-        try:
-            read_index_tar(full_pkg_path, Config())
-        except tarfile.ReadError as re:
-            msg = "tarfile.ReadError encountered. Original error: {}".format(re)
-            msg += "\nRemoving bad package: {}".format(full_pkg_path)
-            logger.error(msg)
-            yield full_pkg_path
-
-
-def _remove_package(pkg_path):
-    """
-    Log and remove a package.
-
-    Parameters
-    ----------
-    pkg_path : str
-        Path to a conda package that should be removed
-    """
-    logger.info("Removing: {}".format(pkg_path))
-    os.remove(pkg_path)
+    logger.info(pformat(sorted(to_mirror)))
 
+    # 6. for each download:
+    # a. download to temp file
+    # b. validate contents of temp file
+    # c. move to local repo
+    # mirror all new packages
+    with tempfile.TemporaryDirectory() as download_dir:
+        logging.info('downloading to the tempdir %s', download_dir)
+        for package_name in sorted(to_mirror):
+            url = DOWNLOAD_URL.format(
+                channel=upstream_channel,
+                name=repodata[package_name]['name'],
+                version=repodata[package_name]['version'],
+                platform=platform,
+                file_name=package_name)
+            _download(url, download_dir, repodata)
+
+        # 8. download repodata.json and repodata.json.bz2
+        url = REPODATA.format(channel=upstream_channel, platform=platform)
+        _download(url, download_dir, validate=False)
+        url = REPODATA.format(channel=upstream_channel, platform=platform) + ".bz2"
+        _download(url, download_dir, validate=False)
+
+        # validate all packages in the download directory
+        _validate_packages(repodata, download_dir)
+        logging.debug('contents of %s are %s',
+                      download_dir,
+                      pformat(os.listdir(download_dir)))
+        for f in os.listdir(download_dir):
+
+            if f.endswith('json') or f.endswith('.bz2'):
```

Comment:
Maybe I'm wrong, but I expect most changes to be addition of new packages rather than updates. In that case, it seems that moving all of the packages into place and then the JSON file would reduce the window of time when a user might do a `conda install foo` and get an error about a missing package because it hasn't been copied into place yet.